gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java

```diff
@@ -0,0 +1,181 @@
+    
+    if(props.containsKey(CONFIG_STORE_REPLICATION_DISABLE_TAGS)){
+      List<String> disableStrs = Splitter.on(",").omitEmptyStrings().splitToList(props.getProperty(CONFIG_STORE_REPLICATION_DISABLE_TAGS));
+      List<Path> disablePaths = new ArrayList<Path>();
+      for(String s: disableStrs){
+        disablePaths.add(PathUtils.mergePaths(new Path(this.storeRoot), new Path(s)));
+      }
+      this.replicationDisableTags = Optional.of(disablePaths);
+    }
+    else{
+      this.replicationDisableTags = Optional.absent();
+    }
+
+    configClient = ConfigClient.createConfigClient(VersionStabilityPolicy.WEAK_LOCAL_STABILITY);
+  }
+
+  protected static Set<URI> getValidDatasetURIs(Collection<URI> allDatasetURIs, Collection<URI> disabledURIs, Path ancestor) {
+    if (allDatasetURIs == null || allDatasetURIs.isEmpty()) {
+      return ImmutableSet.of();
+    }
+
+    Comparator<URI> pathLengthComparator = new Comparator<URI>() {
+      public int compare(URI c1, URI c2) {
+        return c1.getPath().length() - c2.getPath().length();
+      }
+    };
+
+    List<URI> datasetsList = new ArrayList<URI>(allDatasetURIs);
+
+    // sort the URI based on the path length to make sure the parent path appear before children
+    Collections.sort(datasetsList, pathLengthComparator);
+
+    Set<URI> disabledSet = new HashSet<URI>(disabledURIs);
+    Set<URI> validURISet = new HashSet<URI>();
+
+    // remove non leaf URI
+    for (URI u : datasetsList) {
+      URI needToRemove = null;
+
+      // valid dataset URI
+      if (PathUtils.isAncestor(ancestor, new Path(u.getPath()))) {
+        
+        for (URI leaf : validURISet) {
+          if (PathUtils.isAncestor(new Path(leaf.getPath()), new Path(u.getPath()))) {
+            // due to the sort, at most one element need to be removed from leafDatasets
+            needToRemove = leaf;
+            break;
+          }
+        }
+
+        if (needToRemove != null) {
+          validURISet.remove(needToRemove);
+        }
+        validURISet.add(u);
+      }
+    }
+    
+    // remove disabled URIs
+    for(URI disable: disabledSet){
+      if (validURISet.remove(disable) ){
+        log.info("skip disabled dataset " + disable );
+      }
+    }
+    return validURISet;
+  }
+
+  /**
+   * Based on the {@link #replicationTag}, find all URI which imports the tag. Then filter out
+   * 
+   * 1. disabled dataset URI
+   * 2. None leaf dataset URI
+   * 
+   * Then created {@link ConfigBasedDataset} based on the {@link Config} of the URIs
+   */
+  @Override
+  public List<ConfigBasedDataset> findDatasets() throws IOException {
+    Collection<URI> allDatasetURIs;
+    Collection<URI> disabledURIs = ImmutableList.of();;
+    try {
+      // get all the URIs which imports {@link #replicationTag}
+      allDatasetURIs = configClient.getImportedBy(new URI(replicationTag.toString()), true);
+      
+      if(this.replicationDisableTags.isPresent()){
+        disabledURIs = new ArrayList<URI>();
```

Comment:
You seem to first initialize this as a list and then you convert to a set. You might as well, use a set to begin with.